google/resource_composer_environment.go

```diff
@@ -0,0 +1,974 @@
+		var err error
+		project, err = getProject(d, config)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Check if zone was given in node_config.
+	if nodeCfgZone != nil {
+		z := GetResourceNameFromSelfLink(nodeCfgZone.(string))
+		if len(z) > 0 {
+			if len(zone) > 0 && zone != z {
+				// Make sure machine type and zone locations do not conflict.
+				return "", fmt.Errorf(`invalid machine_type: "%s" for specified zone "%s" in node_config`, machineType, zone)
+			}
+			zone = z
+		}
+	}
+	if len(zone) == 0 {
+		return "", fmt.Errorf(`zone must be specified for machine_type "%s" as separate zone field or in machine_type`, machineType)
+	}
+	if len(machineType) == 0 {
+		return "", fmt.Errorf(`invalid machine_type "%s"`, fullMachineType)
+	}
+
+	return fmt.Sprintf("projects/%s/zones/%s/machineTypes/%s", project, zone, machineType), nil
+}
+
+func expandComposerEnvironmentNetwork(v interface{}, d *schema.ResourceData, config *Config) (string, error) {
+	network := v.(string)
+	if len(network) == 0 {
+		return "", nil
+	}
+
+	r := regexp.MustCompile("projects/(.+)/networks/(.+)")
+	if r.MatchString(network) {
+		return getRelativePath(network)
+	}
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("projects/%s/global/networks/%s", project, network), nil
+}
+
+func expandComposerEnvironmentSubnetwork(v interface{}, d *schema.ResourceData, config *Config) (string, error) {
+	subnet := v.(string)
+	if len(subnet) == 0 {
+		return "", nil
+	}
+
+	r := regexp.MustCompile("projects/(.+)/regions/(.+)/subnetworks/(.+)")
+	if r.MatchString(subnet) {
+		return getRelativePath(subnet)
+	}
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return "", err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("projects/%s/regions/%s/subnetworks/%s", project, region, subnet), nil
+}
+
+func expandComposerEnvironmentSetList(v interface{}, d *schema.ResourceData, config *Config) ([]string, error) {
+	if v == nil {
+		return nil, nil
+	}
+	return convertStringArr(v.(*schema.Set).List()), nil
+}
+
+func expandComposerEnvironmentConfigSoftwareConfig(v interface{}, d *schema.ResourceData, config *Config) (*composer.SoftwareConfig, error) {
+	l := v.([]interface{})
+	if len(l) == 0 {
+		return nil, nil
+	}
+	raw := l[0]
+	original := raw.(map[string]interface{})
+	transformed := &composer.SoftwareConfig{}
+
+	transformed.ImageVersion = original["image_version"].(string)
+	transformed.AirflowConfigOverrides = expandComposerEnvironmentStringMap(original["airflow_config_overrides"], d, config)
+	transformed.PypiPackages = expandComposerEnvironmentStringMap(original["pypi_packages"], d, config)
+	transformed.EnvVariables = expandComposerEnvironmentStringMap(original["env_variables"], d, config)
+
+	return transformed, nil
+}
+
+func expandComposerEnvironmentStringMap(v interface{}, d *schema.ResourceData, config *Config) map[string]string {
```

Comment:
I switched it to this, then switched it back because I realized that it needs to get the string map from original. Does d.GetOk("config.0.software_config.0.airflow_config_overrides") work?